File: Python/maximalRectangle.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: maximalRectangle.py
@Function: 最大矩形
@Link: https://leetcode-cn.com/problems/maximal-rectangle/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-14
"""

from typing import List


class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        m, n = len(matrix), len(matrix[0])
        # 记录当前位置上方连续“1”的个数
        pre = [0]*(n+1)
        res = 0
        for i in range(m):
            for j in range(n):
                # 前缀和
                pre[j] = pre[j]+1 if matrix[i][j] == "1" else 0

            # 单调栈
            stack = [-1]
            for k, num in enumerate(pre):
                while stack and pre[stack[-1]] > num:
                    index = stack.pop()
                    res = max(res, pre[index]*(k-stack[-1]-1))
                stack.append(k)

        return res


# ----------------------------------
if __name__ == "__main__":
    matrix = [["1", "0", "1", "0", "0"], ["1", "0", "1", "1", "1"],
              ["1", "1", "1", "1", "1"], ["1", "0", "0", "1", "0"]]

    solution = Solution()
    max_area_maxtrix = solution.maximalRectangle(matrix)
    print(f"The solution of this problem is {max_area_maxtrix}")

File: C++/maximalRectangle.cpp
/**
 * @File    : maximalRectangle.cpp
 * @Brief   : 最大矩形
 * @Link    : https://leetcode-cn.com/problems/maximal-rectangle/
 * @Author  : Wei Li
 * @Date    : 2021-08-14
*/

#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>

class Solution
{
public:
    int maximalRectangle(std::vector<std::vector<char>> &matrix)
    {
        int m = matrix.size();
        if (m == 0)
        {
            return 0;
        }
        int n = matrix[0].size();
        std::vector<std::vector<int>> left(m, std::vector<int>(n, 0));

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (matrix[i][j] == '1')
                {
                    left[i][j] = (j == 0 ? 0 : left[i][j - 1]) + 1;
                }
            }
        }

        int ret = 0;
        for (int j = 0; j < n; j++)
        { // 对于每一列，使用基于柱状图的方法
            std::vector<int> up(m, 0), down(m, 0);

            std::stack<int> stk;
            for (int i = 0; i < m; i++)
            {
                while (!stk.empty() && left[stk.top()][j] >= left[i][j])
                {
                    stk.pop();
                }
                up[i] = stk.empty() ? -1 : stk.top();
                stk.push(i);
            }
            stk = std::stack<int>();
            for (int i = m - 1; i >= 0; i--)
            {
                while (!stk.empty() && left[stk.top()][j] >= left[i][j])
                {
                    stk.pop();
                }
                down[i] = stk.empty() ? m : stk.top();
                stk.push(i);
            }

            for (int i = 0; i < m; i++)
            {
                int height = down[i] - up[i] - 1;
                int area = height * left[i][j];
                ret = std::max(ret, area);
            }
        }
        return ret;
    }
};

// --------------------------------
int main(int argc, char **argv)
{
    std::vector<std::vector<char>> matrix{{'1', '0', '1', '0', '0'}, {'1', '0', '1', '1', '1'}, {'1', '1', '1', '1', '1'}, {'1', '0', '0', '1', '0'}};

    Solution solution;
    auto max_area_matrix = solution.maximalRectangle(matrix);
    std::cout << "The solution of this problem is " << max_area_matrix << std::endl;

    return 0;
}
